presto-main/src/main/java/io/prestosql/server/security/oauth2/OAuth2Authenticator.java

```diff
@@ -38,12 +55,62 @@ public OAuth2Authenticator(OAuth2Service service)
     public Identity authenticate(ContainerRequestContext request)
             throws RedirectAuthenticationException
     {
-        Challenge.Started challenge = service.startChallenge();
-        throw new RedirectAuthenticationException(
-                "Unauthorized",
-                format("Bearer realm=\"Presto\", authorizationUrl=\"%s\", status=\"%s\"",
-                        challenge.getAuthorizationUrl(),
-                        challenge.getStatus()),
-                challenge.getAuthorizationUrl());
+        return getAccessToken(request)
+                .map(token -> Identity.forUser(token.getBody().getSubject()).build())
+                .orElseThrow(() -> {
+                    Challenge.Started challenge = service.startChallenge();
+                    return new RedirectAuthenticationException(
+                            "Unauthorized",
+                            format("Bearer realm=\"Presto\", authorizationUrl=\"%s\", status=\"%s\"",
+                                    challenge.getAuthorizationUrl(),
+                                    challenge.getStatus()),
+                            challenge.getAuthorizationUrl());
+                });
+    }
+
+    private Optional<Jws<Claims>> getAccessToken(ContainerRequestContext request)
+    {
+        Stream<String> accessTokenSources = Stream.concat(Stream.concat(
+                getTokenFromCookie(request),
+                getTokenFromHeader(request)),
+                getTokenFromQueryParam(request));
+        return accessTokenSources
+                .filter(not(String::isBlank))
+                .map(token -> {
+                    try {
+                        return Optional.ofNullable(service.parseClaimsJws(token));
+                    }
+                    catch (JwtException | IllegalArgumentException | UncheckedJwkException e) {
+                        LOG.debug("Unable to parse JWT token: " + e.getMessage(), e);
+                        return Optional.<Jws<Claims>>empty();
+                    }
+                })
+                .findFirst()
+                .flatMap(Function.identity());
+    }
+
+    private Stream<String> getTokenFromCookie(ContainerRequestContext request)
+    {
+        return Stream
+                .ofNullable(request.getCookies().get(PRESTO_UI_COOKIE))
+                .map(Cookie::getValue);
+    }
+
+    private Stream<String> getTokenFromHeader(ContainerRequestContext request)
+    {
+        return Stream
+                .ofNullable(request.getHeaders().get(AUTHORIZATION))
+                .flatMap(Collection::stream)
+                .filter(Objects::nonNull)
+                .filter(header -> header.startsWith("Bearer "))
+                .map(header -> header.replaceFirst("^Bearer ", ""));
```

Comment:
This will compile the regular expression for each call.  Instead just do something like `header.substring("Bearer ".length())`